schematools/permissions.py

```diff
@@ -0,0 +1,107 @@
+
+
+def revoke_permissions(engine, role):
+    grantee =role
+    schema_relation_infolist = query.get_all_table_acls(engine, schema='public')
+    for schema_relation_info in schema_relation_infolist:
+        if schema_relation_info.acl:
+            acl_list = [parse_acl_item(item) for item in schema_relation_info.acl]
+            for acl in acl_list:
+                if acl.grantee == role:
+                    print('revoking ALL priviliges of role "{}" on table "{}"'.format(role, schema_relation_info.name))
+                    revoke_statement = revoke("ALL", PgObjectType.TABLE, schema_relation_info.name, grantee)
+                    engine.execute(revoke_statement)
+
+def apply_schema_and_profile_permissions(engine, ams_schema, profiles, role, scope):
+    if ams_schema:
+        create_acl_from_schemas(engine, ams_schema, role, scope)
+    if profiles:
+        profile_list = profiles.values()
+        create_acl_from_profiles(engine, 'public', profile_list, role, scope)
+
+def create_acl_from_profiles(engine, schema, profile_list, role, scope):
+    acl_list = query.get_all_table_acls(engine, schema='public')
+    priviliges = ["SELECT", ]
+    grantee = role
+    for profile in profile_list:
+        if scope in profile["scopes"]:
+            for dataset, details in profile["schema_data"]["datasets"].items():
+                for item in acl_list:
+                    if item.name.startswith(dataset+"_"):
+                        grant_statement = grant(priviliges, PgObjectType.TABLE, item.name, grantee, grant_option=False, schema=schema)
+                        print(grant_statement)
+                        engine.execute(grant_statement)
+
+
+def create_acl_from_schema(engine, ams_schema, role, scope):
+    grantee = role
+    dataset_scope = ams_schema.auth if ams_schema.auth else {PUBLIC_SCOPE, }
+    dataset_scope_set = {dataset_scope} if isinstance(dataset_scope, str) else set(dataset_scope)
+
+    if len(dataset_scope_set) > 1 or PUBLIC_SCOPE not in dataset_scope:
+        print('Found dataset read permission for "{}" to scopes "{}"'.format(ams_schema.id, dataset_scope_set))
+    for table in ams_schema.get_tables(include_nested=True, include_through=True):
+        table_name = "{}_{}".format(table.dataset.id, to_snake_case(table.id))  # een aantal table.id's zijn camelcase
+        table_scope = table.auth if table.auth else dataset_scope
+        table_scope_set = {table_scope} if isinstance(table_scope, str) else set(table_scope)
+        if table.auth:
+            print('Found table read permission for "{}" to scopes "{}"'.format(table_name, table_scope_set))
+            print('"{}" overrules "{}" for read permission of "{}"'.format(table_scope_set, dataset_scope_set, table_name))
+        contains_field_grants = False
+        fields = [field for field in table.fields if '$ref' not in field]
```

Comment:
is er geen grant mogelijk op een `..._id` veld? misschien even comment erboven waarom dat bewust zo gedaan wordt.

Reply:
Dit is idd een fout. Goed gevonden. Dit deed ik om het schema referentie veld uit te sluiten, maar zo mis ik ook .._id. Dit ga ik fixen.

Reply:
Een van de table.fields is `schema`, maar dat veld vertaalt uiteindelijk niet naar een db column, het is slechts een verwijzing naar het schema url. Ik filter dat veld er nu uit, maar beter zou zijn als het niet in table.fields zou zitten. Iets om later uit te zoeken/fixen. Ik filter voor nu simpelweg `schema` eruit.